## Importing required libraries

In [1]:
import pandas as pd

## Loading the dataset

In [2]:
df = pd.read_csv('AB_NYC_2019.csv')

## Number of rows and columns

In [3]:
df.shape

(48895, 16)

## Access first five rows

In [4]:
df.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

## Displaying all column names from dataset 

In [5]:
df.columns

Index(['id', 'name', 'host_id', 'host_name', 'neighbourhood_group',
       'neighbourhood', 'latitude', 'longitude', 'room_type', 'price',
       'minimum_nights', 'number_of_reviews', 'last_review',
       'reviews_per_month', 'calculated_host_listings_count',
       'availability_365'],
      dtype='object')

# 1. Data Integrity

## Check column types

In [6]:
df.dtypes

id                                  int64
name                               object
host_id                             int64
host_name                          object
neighbourhood_group                object
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
price                               int64
minimum_nights                      int64
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
dtype: object

## DataFrame information

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

From above information columns like name, host_name, last_review, reviews_per_month have missing values which needed to analysis further is required


## Checking missing values for each column

In [8]:
df.isna()

id   name  host_id  host_name  neighbourhood_group  neighbourhood  \
0      False  False    False      False                False          False   
1      False  False    False      False                False          False   
2      False  False    False      False                False          False   
3      False  False    False      False                False          False   
4      False  False    False      False                False          False   
...      ...    ...      ...        ...                  ...            ...   
48890  False  False    False      False                False          False   
48891  False  False    False      False                False          False   
48892  False  False    False      False                False          False   
48893  False  False    False      False                False          False   
48894  False  False    False      False                False          False   

       latitude  longitude  room_type  price  minimum_nights  \
0         False      False      False  False           False   
1         False      False      False  False           False   
2         False      False      False  False           False   
3         False      False      False  False           False   
4         False      False      False  False           False   
...         ...        ...        ...    ...             ...   
48890     False      False      False  False           False   
48891     False      False      False  False           False   
48892     False      False      False  False           False   
48893     False      False      False  False           False   
48894     False      False      False  False           False   

       number_of_reviews  last_review  reviews_per_month  \
0                  False        False              False   
1                  False        False              False   
2                  False         True               True   
3                  False        False              False   
4                  False        False              False   
...                  ...          ...                ...   
48890              False         True               True   
48891              False         True               True   
48892              False         True               True   
48893              False         True               True   
48894              False         True               True   

       calculated_host_listings_count  availability_365  
0                               False             False  
1                               False             False  
2                               False             False  
3                               False             False  
4                               False             False  
...                               ...               ...  
48890                           False             False  
48891                           False             False  
48892                           False             False  
48893                           False             False  
48894                           False             False  

[48895 rows x 16 columns]

## Checking missing values count for each column

In [9]:
df.isna().sum()


id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

## Checking missing values count for each column after sorting

In [10]:
df.isna().sum().sort_values(ascending = False)

last_review                       10052
reviews_per_month                 10052
host_name                            21
name                                 16
id                                    0
host_id                               0
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
calculated_host_listings_count        0
availability_365                      0
dtype: int64

## Checking for duplicated rows in the dataset

In [11]:
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
48890    False
48891    False
48892    False
48893    False
48894    False
Length: 48895, dtype: bool

## Checking how many duplicated rows in the dataset

In [12]:
df.duplicated().sum() # sum(df.duplicated())

0

This means the dataset maintains its integrity in terms of unique entries.

## Checking for negative values in the dataset

In [13]:
df[df.price < 0]

Empty DataFrame
Columns: [id, name, host_id, host_name, neighbourhood_group, neighbourhood, latitude, longitude, room_type, price, minimum_nights, number_of_reviews, last_review, reviews_per_month, calculated_host_listings_count, availability_365]
Index: []

This means the dataset doesn't contain any negative values

# 2. Missing Data Handling

## Check how many missing values are in the dataset

In [14]:
df.isna().sum().sort_values(ascending = False)

last_review                       10052
reviews_per_month                 10052
host_name                            21
name                                 16
id                                    0
host_id                               0
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
calculated_host_listings_count        0
availability_365                      0
dtype: int64

## Dropping columns with significant missing data 

In [15]:
df_new = df.drop(columns = ['name', 'host_name', 'last_review'])

## Checking the missing values count of new dataframe 

In [16]:
df_new.isna().sum()

id                                    0
host_id                               0
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

Since dropped columns are not crucial for the analysis and their missing data could affect the overall quality of the dataset

## Checking the average value of reviews_per_month

In [17]:
reviews_per_month_avg = df_new['reviews_per_month'].mean().round(2)
reviews_per_month_avg

1.37

## Filling the missing rows in reviews_per_month column

In [18]:
df_new.fillna(reviews_per_month_avg, inplace = True)

Since filled rows are crucial for the analysis and their data could decide the overall quality of the dataset

## Checking missing values in the new dataframe

In [19]:
df_new.isna().sum()

id                                0
host_id                           0
neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
dtype: int64

The dataset is now free from missing entries, and it's ready for further analysis or modeling.

# 3. Duplicate Removal

## Check for duplicate rows in the new dataframe

In [20]:
df_new.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
48890    False
48891    False
48892    False
48893    False
48894    False
Length: 48895, dtype: bool

## Count duplicates rows 

In [21]:
df_new.duplicated().sum()

0

There are no duplicate rows in the dataframe and no further action is needed for duplicate removal.

# 4. Standardization

## Access the first five rows

In [22]:
df_new.head()

id  host_id neighbourhood_group neighbourhood  latitude  longitude  \
0  2539     2787            Brooklyn    Kensington  40.64749  -73.97237   
1  2595     2845           Manhattan       Midtown  40.75362  -73.98377   
2  3647     4632           Manhattan        Harlem  40.80902  -73.94190   
3  3831     4869            Brooklyn  Clinton Hill  40.68514  -73.95976   
4  5022     7192           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews  \
0     Private room    149               1                  9   
1  Entire home/apt    225               1                 45   
2     Private room    150               3                  0   
3  Entire home/apt     89               1                270   
4  Entire home/apt     80              10                  9   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2               1.37                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

## Convert the all text columns into lowercase

In [23]:
text_columns = ['neighbourhood_group','neighbourhood', 'room_type']
for column in text_columns:
    df_new[column] = df_new[column].str.lower()

## After conversion check the new text columns

In [24]:
df_new.head()

id  host_id neighbourhood_group neighbourhood  latitude  longitude  \
0  2539     2787            brooklyn    kensington  40.64749  -73.97237   
1  2595     2845           manhattan       midtown  40.75362  -73.98377   
2  3647     4632           manhattan        harlem  40.80902  -73.94190   
3  3831     4869            brooklyn  clinton hill  40.68514  -73.95976   
4  5022     7192           manhattan   east harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews  \
0     private room    149               1                  9   
1  entire home/apt    225               1                 45   
2     private room    150               3                  0   
3  entire home/apt     89               1                270   
4  entire home/apt     80              10                  9   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2               1.37                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

This step ensures uniformity across the dataframe, making it easier to analyze and interpret the data accurately.

# 5. Outlier Detection

We'll calculate the Z-score for each numerical columns. Any value with a Z-score above a certain threshold (commonly 3 or below -3) will be considered an outlier.

## Access the dataframe to identify numerical columns

In [25]:
df_new.head()

id  host_id neighbourhood_group neighbourhood  latitude  longitude  \
0  2539     2787            brooklyn    kensington  40.64749  -73.97237   
1  2595     2845           manhattan       midtown  40.75362  -73.98377   
2  3647     4632           manhattan        harlem  40.80902  -73.94190   
3  3831     4869            brooklyn  clinton hill  40.68514  -73.95976   
4  5022     7192           manhattan   east harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews  \
0     private room    149               1                  9   
1  entire home/apt    225               1                 45   
2     private room    150               3                  0   
3  entire home/apt     89               1                270   
4  entire home/apt     80              10                  9   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2               1.37                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

## Import required library for Z-score

In [26]:
from scipy.stats import zscore

## Calculate the Z-score for numerical columns

In [27]:
numerical_columns = ['price', 'minimum_nights',	'number_of_reviews', 'reviews_per_month', 'calculated_host_listings_count',	'availability_365']
z_scores = zscore(df_new[numerical_columns])
z_scores

price  minimum_nights  number_of_reviews  reviews_per_month  \
0     -0.015493       -0.293996          -0.320414          -0.776199   
1      0.300974       -0.293996           0.487665          -0.662696   
2     -0.011329       -0.196484          -0.522433          -0.001709   
3     -0.265335       -0.293996           5.538156           2.181552   
4     -0.302811        0.144807          -0.320414          -0.849642   
...         ...             ...                ...                ...   
48890 -0.344452       -0.245240          -0.522433          -0.001709   
48891 -0.469373       -0.147729          -0.522433          -0.001709   
48892 -0.157070        0.144807          -0.522433          -0.001709   
48893 -0.406912       -0.293996          -0.522433          -0.001709   
48894 -0.261171       -0.001461          -0.522433          -0.001709   

       calculated_host_listings_count  availability_365  
0                           -0.034716          1.916250  
1                           -0.156104          1.840275  
2                           -0.186451          1.916250  
3                           -0.186451          0.617065  
4                           -0.186451         -0.856865  
...                               ...               ...  
48890                       -0.156104         -0.788486  
48891                       -0.156104         -0.583352  
48892                       -0.186451         -0.651730  
48893                       -0.034716         -0.841669  
48894                       -0.186451         -0.682120  

[48895 rows x 6 columns]

## Set the threshold for Z-score 

In [28]:
threshold = 3

## Filter rows where the Z-score is greater than the threshold

In [29]:
outliers = (z_scores > threshold).any(axis = 1)
df_new[outliers]

id    host_id neighbourhood_group   neighbourhood  latitude  \
3          3831       4869            brooklyn    clinton hill  40.68514   
7          5178       8967           manhattan  hell's kitchen  40.76489   
9          5238       7549           manhattan       chinatown  40.71344   
11         5441       7989           manhattan  hell's kitchen  40.76076   
12         5803       9744            brooklyn     south slope  40.66829   
...         ...        ...                 ...             ...       ...   
48723  36404684  107434423           manhattan          nolita  40.72283   
48724  36404784  107434423           manhattan  hell's kitchen  40.76082   
48725  36404815  107434423           manhattan  hell's kitchen  40.76083   
48726  36404936  107434423            brooklyn    williamsburg  40.71493   
48727  36404972  107434423           manhattan    east village  40.73206   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
3      -73.95976  entire home/apt     89               1                270   
7      -73.98493     private room     79               2                430   
9      -73.99037  entire home/apt    150               1                160   
11     -73.98867     private room     85               2                188   
12     -73.98779     private room     89               4                167   
...          ...              ...    ...             ...                ...   
48723  -73.99472  entire home/apt    316              30                  0   
48724  -73.99709  entire home/apt    385              30                  0   
48725  -73.99727  entire home/apt    267              30                  0   
48726  -73.96365  entire home/apt    278              30                  0   
48727  -73.98776  entire home/apt    365              30                  0   

       reviews_per_month  calculated_host_listings_count  availability_365  
3                   4.64                               1               194  
7                   3.47                               1               220  
9                   1.33                               4               188  
11                  1.50                               1                39  
12                  1.34                               3               314  
...                  ...                             ...               ...  
48723               1.37                             232               325  
48724               1.37                             232               338  
48725               1.37                             232               228  
48726               1.37                             232               188  
48727               1.37                             232               295  

[3263 rows x 13 columns]

There are 3263 outliers exist which are significantly very far the mean value of the numerical columns in the dataframe

## Drop the outliers in the dataframe

In [30]:
outliers_index = df_new[outliers].index
df_cleaned = df_new.drop(index = outliers_index)

## Check the dataframe with outliers and without outliers

In [31]:
print('Dataframe with outliers:', df_new.shape)
print('Dataframe without outliers:', df_cleaned.shape)

Dataframe with outliers: (48895, 13)
Dataframe without outliers: (45632, 13)


# Check the cleaned dataframe

In [32]:
df_cleaned

id   host_id neighbourhood_group       neighbourhood  latitude  \
0          2539      2787            brooklyn          kensington  40.64749   
1          2595      2845           manhattan             midtown  40.75362   
2          3647      4632           manhattan              harlem  40.80902   
4          5022      7192           manhattan         east harlem  40.79851   
5          5099      7322           manhattan         murray hill  40.74767   
...         ...       ...                 ...                 ...       ...   
48890  36484665   8232441            brooklyn  bedford-stuyvesant  40.67853   
48891  36485057   6570630            brooklyn            bushwick  40.70184   
48892  36485431  23492952           manhattan              harlem  40.81475   
48893  36485609  30985759           manhattan      hell's kitchen  40.75751   
48894  36487245  68119814           manhattan      hell's kitchen  40.76404   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0      -73.97237     private room    149               1                  9   
1      -73.98377  entire home/apt    225               1                 45   
2      -73.94190     private room    150               3                  0   
4      -73.94399  entire home/apt     80              10                  9   
5      -73.97500  entire home/apt    200               3                 74   
...          ...              ...    ...             ...                ...   
48890  -73.94995     private room     70               2                  0   
48891  -73.93317     private room     40               4                  0   
48892  -73.94867  entire home/apt    115              10                  0   
48893  -73.99112      shared room     55               1                  0   
48894  -73.98933     private room     90               7                  0   

       reviews_per_month  calculated_host_listings_count  availability_365  
0                   0.21                               6               365  
1                   0.38                               2               355  
2                   1.37                               1               365  
4                   0.10                               1                 0  
5                   0.59                               1               129  
...                  ...                             ...               ...  
48890               1.37                               2                 9  
48891               1.37                               2                36  
48892               1.37                               1                27  
48893               1.37                               6                 2  
48894               1.37                               1                23  

[45632 rows x 13 columns]

## Reindexing the cleaned dataframe

In [33]:
df_cleaned.reset_index(drop = True, inplace = True)

## Achieved cleaned dataframe from the dataset

In [34]:
df_cleaned

id   host_id neighbourhood_group       neighbourhood  latitude  \
0          2539      2787            brooklyn          kensington  40.64749   
1          2595      2845           manhattan             midtown  40.75362   
2          3647      4632           manhattan              harlem  40.80902   
3          5022      7192           manhattan         east harlem  40.79851   
4          5099      7322           manhattan         murray hill  40.74767   
...         ...       ...                 ...                 ...       ...   
45627  36484665   8232441            brooklyn  bedford-stuyvesant  40.67853   
45628  36485057   6570630            brooklyn            bushwick  40.70184   
45629  36485431  23492952           manhattan              harlem  40.81475   
45630  36485609  30985759           manhattan      hell's kitchen  40.75751   
45631  36487245  68119814           manhattan      hell's kitchen  40.76404   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0      -73.97237     private room    149               1                  9   
1      -73.98377  entire home/apt    225               1                 45   
2      -73.94190     private room    150               3                  0   
3      -73.94399  entire home/apt     80              10                  9   
4      -73.97500  entire home/apt    200               3                 74   
...          ...              ...    ...             ...                ...   
45627  -73.94995     private room     70               2                  0   
45628  -73.93317     private room     40               4                  0   
45629  -73.94867  entire home/apt    115              10                  0   
45630  -73.99112      shared room     55               1                  0   
45631  -73.98933     private room     90               7                  0   

       reviews_per_month  calculated_host_listings_count  availability_365  
0                   0.21                               6               365  
1                   0.38                               2               355  
2                   1.37                               1               365  
3                   0.10                               1                 0  
4                   0.59                               1               129  
...                  ...                             ...               ...  
45627               1.37                               2                 9  
45628               1.37                               2                36  
45629               1.37                               1                27  
45630               1.37                               6                 2  
45631               1.37                               1                23  

[45632 rows x 13 columns]

The task of Data Cleaning has been successfully completed, ensuring a well-prepared and reliable dataset that will support smooth and accurate analysis moving forward. 